worlds/checksmate/__init__.py

```diff
@@ -0,0 +1,386 @@
+        "multiworld, and related software.",
+        "English",
+        "checksmate_en.md",
+        "checks-mate/en",
+        ["roty", "rft50"]
+    )]
+
+
+class CMWorld(World):
+    """
+    ChecksMate is a game where you play chess, but all of your pieces were scattered across the multiworld.
+    You win when you checkmate the opposing king!
+    """
+    game: ClassVar[str] = "ChecksMate"
+    option_definitions = cm_options
+    data_version = 0
+    web = CMWeb()
+    required_client_version = (0, 3, 4)
+
+    item_name_to_id = {name: data.code for name, data in item_table.items()}
+    location_name_to_id = {name: data.code for name, data in location_table.items()}
+    locked_locations: List[str]
+
+    item_name_groups = {
+        # "Pawn": {"Pawn A", "Pawn B", "Pawn C", "Pawn D", "Pawn E", "Pawn F", "Pawn G", "Pawn H"},
+        "Enemy Pawn": {"Enemy Pawn A", "Enemy Pawn B", "Enemy Pawn C", "Enemy Pawn D",
+                       "Enemy Pawn E", "Enemy Pawn F", "Enemy Pawn G", "Enemy Pawn H"},
+        "Enemy Piece": {"Enemy Piece A", "Enemy Piece B", "Enemy Piece C", "Enemy Piece D",
+                        "Enemy Piece F", "Enemy Piece G", "Enemy Piece H"},
+    }
+    items_used: Dict[int, Dict[str, int]] = {}
+    army: Dict[int, int] = {}
+
+    item_pool: List[CMItem] = []
+    prefill_items: List[CMItem] = []
+
+    known_pieces = {"Progressive Minor Piece": 9, "Progressive Major Piece": 5, "Progressive Major To Queen": 4, }
+    piece_type_limit_options = {
+        "Progressive Minor Piece": "minor_piece_limit_by_type",
+        "Progressive Major Piece": "major_piece_limit_by_type",
+        "Progressive Major To Queen": "queen_piece_limit_by_type",
+    }
+    piece_limit_options = {
+        "Progressive Major To Queen": "queen_piece_limit",
+    }
+    piece_types_by_army = {
+        # Vanilla
+        0: {"Progressive Minor Piece": 2, "Progressive Major Piece": 1, "Progressive Major To Queen": 1},
+        # Colorbound Clobberers (the War Elephant is rather powerful)
+        1: {"Progressive Minor Piece": 1, "Progressive Major Piece": 2, "Progressive Major To Queen": 1},
+        # Remarkable Rookies
+        2: {"Progressive Minor Piece": 2, "Progressive Major Piece": 1, "Progressive Major To Queen": 1},
+        # Nutty Knights (although the Short Rook and Half Duck swap potency)
+        3: {"Progressive Minor Piece": 2, "Progressive Major Piece": 1, "Progressive Major To Queen": 1},
+        # Eurasian pieces
+        4: {"Progressive Minor Piece": 2, "Progressive Major Piece": 1, "Progressive Major To Queen": 1},
+    }
+
+    def __init__(self, multiworld: MultiWorld, player: int):
+        super(CMWorld, self).__init__(multiworld, player)
+        self.locked_locations = []
+
+    def generate_early(self) -> None:
+        army_constraint = get_option_value(self.multiworld, self.player, "fairy_chess_army")
```

Comment:
When swapping to the options API, this turns into `army_constraint = self.options.fairy_chess_army.value`.

Reply:
An item of mine is limited by each of several options. This is one reason to move to an ItemDict, and I would consider that option given advice in favor of it. However, assuming I won't, I need to migrate this to the Options API, and would appreciate advice on the matter.

Maybe I should add this as a property to the given Options, as they are objects where I might reference additional features?

```py
    piece_type_limit_options = {
        "Progressive Minor Piece": "minor_piece_limit_by_type",
        "Progressive Major Piece": "major_piece_limit_by_type",
        "Progressive Major To Queen": "queen_piece_limit_by_type",
    }
...
        if chosen_item in self.piece_limit_options:
            piece_total_limit = self.options[self.piece_limit_options[chosen_item]].value
```

Reply:
I have moved this dictionary to the CMOptions object as a classvar, and changed the values to a lambda which accepts a specific CMOptions.